Add Azure File Services API module

Krajzys opened the pull request:
<!-- Thank you for submitting a Pull Request. If you're new to contributing to BCApps please read our pull request guideline below
* https://github.com/microsoft/BCApps/Contributing.md
-->
# Summary <!-- Provide a general summary of your changes -->

This is a port of the pull request from the old repository, most of the convsersation and reviews are also there (https://github.com/microsoft/ALAppExtensions/pull/23710).
The goal of the PR is to add a module that facilitates working with Azure File Share service (which is a part of the Azure Cloud).
The module is heavily based on the Azure Blob Storage Service API module, and most of the architecture is similair.

There are also tests added for the module, but they should be ran manually, as they require passing credentials.

Fixes #340 

Work item: [AB#477758](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/477758)

pri-kise commented:
We need now Namespaces (and optional using statements)

I'd suggest something like `namespace System.Azure.Storage.Files.Shares;`

Krajzys commented:
Thanks for the tip. How about just `System.Azure.Storage.Files`? Adding shares might be a bit redundant (the only file service on the Azure storage is File Share).

Regarding second point, yes I think we can remove the AFS from most files, but should I then replace it with more descriprive name or just remove the prefix? e.g. AFS Handle -> File Handle.

JesperSchulz commented:
Uh, naming and namespace discussions. I'm curious what @AndreasMoth would suggest 😁

DevHarmonize commented:
@JesperSchulz 

Hi Jesper,
we spoke during Directions Lyon about this PR, Is it still possible to have this in the product in one of the next minors?

kr

JesperSchulz commented:
Security Review scheduled for Tuesday, 12th of December.

JesperSchulz commented:
Security review conducted with @darjoo. Darrick will comment on outcome.

darjoo commented:
After performing the security review with Jesper, the following is the outcome that needs to be fixed.
- Telemetry for all usages

Fx, telemetry when calling `CreateFile`
- OnCall, OnSuccess, OnError

JesperSchulz commented:
@Krajzys, if we get the missing telemetry added next week, we can release this module with 23.3 in January!

Krajzys commented:
@JesperSchulz great! I will get on to it. I should add it through the LogMessage procedure right? I have a questions then. How should I detemine the EventIDs?

Krajzys commented:
I have added the telemetry messages, for now with temporary eventIDs (0000AFS) that need to be changed. Please let me know if the telemetry is enough and how to proceed with the eventIDs 😃.

JesperSchulz commented on src/System Application/Test/DisabledTests/AFSFileClient.json:
```diff
@@ -0,0 +1,8 @@
+[
+    {
+        "bug":  "",
+        "codeunitId":  132517,
```
Why are the tests disabled? Does that mean we've got 0% test coverage?

pri-kise replied:
I think the problem was, when the tests were created, that @darjoo didn't allow any IsHandled pattern for Testing Purpose only.

Therefore those tests require a real Azure File Share Instance for testing.

Azurite doesn't support Azure File Share yet: https://github.com/Azure/Azurite/issues/113

The module maybe needs to be refactored to use the RestClient and give the possibility to specify a HttpClientHanlder.  -> At least for testing purpose. Then it might be possible to Mock Web Requests.

JesperSchulz commented:
> @JesperSchulz great! I will get on to it. I should add it through the LogMessage procedure right? I have a questions then. How should I detemine the EventIDs?

We've got a script which determines the IDs. I'll assign the fitting IDs to your PR! However, we're currently discussing that file names and folder names are customer content and hence cannot get emitted to telemetry. But without that information, what good does telemetry then do? We'll post the outcome of that discussion to this PR. If worst comes to worst, we temporarily revert your last commit and submit the module to the repo in its former state. Then we can create a chaser PR, if we come up with meaningful telemetry!

The 2 things pending for sure though are the disabled tests and the Test Libraries in the Test project. Once those have been fixed, I'm tempted to push the "merge button" 😎

JesperSchulz commented on src/System Application/Test/Azure File Services API/src/AzureTestLibrary.Codeunit.al:
```diff
@@ -0,0 +1,78 @@
+/// <summary>
+/// Provides common authorization functionality for using Azure storage accounts in tests for Azure Storage Services.
+/// </summary>
+codeunit 132516 "Azure Test Library"
```
@grobyns just pointed out, that these probably aren't libraries - in the sense that they would come in handy in other modules too. They are probably rather internal test helpers. Maybe a change of name would suffice here. I'd like to reserve the word Library for proper test libraries, if possible.

darjoo commented:
@Krajzys wonderful job with adding the telemetry. A little too good :D 
Due to privacy reasons, we cannot log any variable potentially containing customer content, EUPI etc etc.
As such, filename, filepath, directory name and all cannot be logged. Please remove those bits.

Krajzys commented:
@darjoo @JesperSchulz Okay, I updated the telemetry messages and renamed the library codeunits to helper.
Regardind the disabled tests. What @pri-kise wrote was correct. We cannot mock requests with current architecture so the module would need to use RestClient. Otherwise we would need to have a regular azure file share setup only for testing purposes and somehow get the credentials for authorizing requests without hardcoding them.

grobyns commented on src/System Application/Test/Azure File Services API/src/AFSTestHelper.Codeunit.al:
```diff
@@ -0,0 +1,56 @@
+
+using System.Azure.Storage;
+
+codeunit 132515 "AFS Test Helper"
```
This name doesn't tell me what the code does... I'd name this 
AFS Get Test Storage Authorization

or a variation/abbreviation thereof... 

also, just FYI, I very much frown on terms like Helper and Management because they do not add anything. You might as well not add them to the name and you'd get the same amount of info... I think we can name things aptly with a little effort.

Krajzys replied:
Okay, I agree. Changed to AFS Get Test Storage Auth.

grobyns commented on src/System Application/Test/Azure File Services API/src/AzureTestHelper.Codeunit.al:
```diff
@@ -0,0 +1,82 @@
+using System.Azure.Storage.Files;
+
+/// <summary>
+/// Provides common authorization functionality for using Azure storage accounts in tests for Azure Storage Services.
```
does it provide common auth functionaliy? As far as i see most of the code clears the fileshare (i.e. a reset or init for tests).

Krajzys replied:
Right, I updated the summary.

grobyns commented on src/System Application/Test/Azure File Services API/src/AzureTestHelper.Codeunit.al:
```diff
@@ -0,0 +1,82 @@
+/// <summary>
+/// Provides common authorization functionality for using Azure storage accounts in tests for Azure Storage Services.
+/// </summary>
+codeunit 132516 "Azure Test Helper"
```
Maybe something like 
AFS Init. Test Storage

?

Krajzys replied:
Yes, that sounds better. Updated.

JesperSchulz commented:
Alright: so here's the plan. Address the remaining comments made by Gert and we'll merge. All HttpClient related code seems to be nicely wrapped in the internal codeunit 8954 "AFS Web Request Helper", which we afterwards can refactor to use the AL Rest Client - which again enables us to better test this.
Does that sound reasonable?

Krajzys commented:
@JesperSchulz Yes, that sounds great. I fixed things pointed out by @grobyns.

JesperSchulz commented:
Awesome! You're REALLY responsive - wow! Let's do this 🥳
Starting CI!

pri-kise commented on src/System Application/App/Azure File Services API/src/Helper/AFSFormatHelper.Codeunit.al:
```diff
@@ -0,0 +1,80 @@
+        exit(DelChr("Value", '=', '{}'));
+    end;
+
+    procedure ConvertToDateTime(PropertyValue: Text): DateTime
```
How will that DateTime Method will work in different languages?

What is the format of DateTime Values, that are returned by the API?
I couldn't find the API Documentation that fast.

Krajzys replied:
This should be language agnostic since it should only be called on the values from the API.
The function you mentioned is used in List Directories calls, and documentation (https://learn.microsoft.com/en-us/rest/api/storageservices/list-directories-and-files#datetime-format-and-api-version-for-timestamp-fields) says it uses the following formats:
ISO8601
RFC1123

In my tests they were evaluated correctly.

JesperSchulz commented:
CI passed! I got a few more comments internally, which I'll address - nothing major. I expect to get this merged before 2024 and in time for the 2023 Update 3 in January.

JesperSchulz commented on src/System Application/App/Azure File Services API/src/Helper/AFSHttpContentHelper.Codeunit.al:
```diff
@@ -0,0 +1,116 @@
+    [NonDebuggable]
+    procedure GetMaxRange(): Integer
+    begin
+        exit(4 * Power(2, 20));
```
What's that number?

pri-kise replied:
I guess this should be maximum ChunkSize for File Upload: 4 MiB -> 4194304 Bytes

Krajzys replied:
Yes, my intention was for the procedure name to define this value, so this would be max range that can be uploaded at once. I can change it to what pri-kise proposed.

JesperSchulz replied:
A comment describing this "magic" number would be fine :-)

JesperSchulz replied:
But I'll take care of that. I'll push a few updates to this PR in the coming days. Nothing major at all.

JesperSchulz commented:
@Krajzys, I have addressed some minor code review comments in my internal 23.3 pull request, which I'd like to push to this branch before merging. Could you give me write access to the branch? When I've pushed my changes, I'd love for you to take one final look, and then we merge. But I've made sure this releases in January 🥳

JesperSchulz commented:
I'll just merge and submit a chaser pack :-)

JesperSchulz commented:
@Krajzys, here a small post-merge request 😋
Any chance you could provide some very simple code samples on how to use this module for the release notes? I could create them myself and get inspired by the tests you've written, but maybe you have some great code already which would show just how simple it is to work with Azure Files?

Krajzys commented:
@JesperSchulz yes, no problem, I could provide some examples. I will try to send them here before the end of today.

Krajzys commented:
Sorry for the delay, I prepared three simple sample pieces of code.
```al
procedure GetFileAsStream(StorageAccount: Text; FileShare: Text; Authorization: Interface "Storage Service Authorization"; FilePath: Text)
var
    AFSFileClient: Codeunit "AFS File Client";
    AFSOperationResponse: Codeunit "AFS Operation Response";
    FileContents: InStream;
begin
    AFSFileClient.Initialize(StorageAccount, FileShare, Authorization);
    AFSOperationResponse := AFSFileClient.GetFileAsStream(FilePath, FileContents);
    if not AFSOperationResponse.IsSuccessful() then
        Error(AFSOperationResponse.GetError());

    DownloadFromStream(FileContents, '', '', '', FilePath);
end;

procedure SendStreamFile(StorageAccount: Text; FileShare: Text; Authorization: Interface "Storage Service Authorization"; FilePath: Text; var FileContent: InStream)
var
    AFSFileClient: Codeunit "AFS File Client";
    AFSOperationResponse: Codeunit "AFS Operation Response";
begin
    AFSFileClient.Initialize(StorageAccount, FileShare, Authorization);
    AFSOperationResponse := AFSFileClient.CreateFile(FilePath, FileContent);
    if not AFSOperationResponse.IsSuccessful() then
        Error(AFSOperationResponse.GetError());

    AFSOperationResponse := AFSFileClient.PutFileStream(FilePath, FileContent);
    if not AFSOperationResponse.IsSuccessful() then
        Error(AFSOperationResponse.GetError());
end;

procedure GetFileNamesInDirectory(StorageAccount: Text; FileShare: Text; Authorization: Interface "Storage Service Authorization"; DirectoryPath: Text[2048]): List of [Text[2048]];
var
    AFSDirectoryContent: Record "AFS Directory Content";
    AFSFileClient: Codeunit "AFS File Client";
    AFSOperationResponse: Codeunit "AFS Operation Response";
    FileNames: List of [Text[2048]];
begin
    AFSFileClient.Initialize(StorageAccount, FileShare, Authorization);
    AFSOperationResponse := AFSFileClient.ListDirectory(DirectoryPath, AFSDirectoryContent);
    if not AFSOperationResponse.IsSuccessful() then
        Error(AFSOperationResponse.GetError());

    AFSDirectoryContent.SetRange("Resource Type", AFSDirectoryContent."Resource Type"::File);
    if not AFSDirectoryContent.FindSet() then
        exit;
    repeat
        FileNames.Add(AFSDirectoryContent."Full Name");
    until AFSDirectoryContent.Next() = 0;

    exit(FileNames);
end;
```

Here is also how I call them:
```al
action(SendFile)
{
    ApplicationArea = All;
    Caption = 'Send File';
    Image = SendAsPDF;
    ToolTip = 'Sends a file to the FileShare';

    trigger OnAction()
    var
        StorageServiceAuthorization: Codeunit "Storage Service Authorization";
        FileStream: InStream;
        FileName: Text;
    begin
        if not UploadIntoStream('', '', '', FileName, FileStream) then
            exit;
        SampleAFSOperations.SendStreamFile(StorageAccount, FileShare, StorageServiceAuthorization.UseReadySAS(SASToken), Path, FileStream);
        Message('File %1 sent succesfully.', FileName);
    end;
}
action(GetFileAsStream)
{
    ApplicationArea = All;
    Caption = 'Get File as Stream';
    ToolTip = 'Gets file as a stream.';
    Image = Download;

    trigger OnAction()
    var
        StorageServiceAuthorization: Codeunit "Storage Service Authorization";
    begin
        SampleAFSOperations.GetFileAsStream(StorageAccount, FileShare, StorageServiceAuthorization.UseReadySAS(SASToken), Path);
    end;
}
```

PJohnstonSysco commented:
@Krajzys  @JesperSchulz  Hi, I'm messing about with this in the most recent release and it works great with SAS Token but when I try to use a Shared Key for authentication I run into issues uploading a file to the File Share. I have attached the error. Any ideas? Authentication using shared key works fine for downloading and listing directory requests.

![image](https://github.com/microsoft/BCApps/assets/123634353/72da983e-db4d-4bdb-88ba-9cbe9bd3785f)

DevHarmonize commented:
I also receive this error when using shared key for upload operation.
But otherwise it works perfectly.

Thanks for your effort in making this @Krajzys !
